src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,412 @@
+
+-- | A string of raw bytes representing a value.
+--
+-- Can also double as a representation for a chunk of a value.
+data RawValue =
+    UnsafeRawValue !Int !Int ByteArray
+  | EmptyRawValue
+  deriving Show
+
+{-# COMPLETE EmptyRawValue, RawValue #-}
+pattern RawValue :: Int -> Int -> ByteArray -> RawValue
+pattern RawValue start end ba <- UnsafeRawValue start end ba
+  where RawValue start end ba =
+          assert (0 <= start && start <= end && end <= sizeofByteArray ba) $
+          UnsafeRawValue start end ba
+
+-- | Size of value in number of bytes.
+sizeofValue :: RawValue -> Int
+sizeofValue EmptyRawValue          = 0
+sizeofValue (RawValue start end _) = end - start
+
+-- | Size of value in number of bytes.
+sizeofValue16 :: RawValue -> Word16
+sizeofValue16 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue32 :: RawValue -> Word32
+sizeofValue32 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue64 :: RawValue -> Word64
+sizeofValue64 = fromIntegral . sizeofValue
+
+-- | TODO: replace this placeholder type by the actual blob reference type.
+data BlobRef = BlobRef {
+    blobRefOffset :: !Word64
+  , blobRefSize   :: !Word32
+  }
+  deriving Show
+
+{-------------------------------------------------------------------------------
+  Page builder
+-------------------------------------------------------------------------------}
+
+-- | A builder representing one or more disk pages.
+--
+-- Typically, this builder only represents a single page, unless it is the
+-- result of adding a larger-than-page value to a run, which made the page
+-- contents exceed the size of a single disk page.
+--
+-- The string of bytes is padded to the target disk-page size (4K).
+--
+-- This builder could be used to serialise into a strict bytestring, lazy
+-- bytestring, or into existing buffers.
+pageBuilder :: PageAcc -> BB.Builder
+pageBuilder PageAcc{..} =
+    -- (1) directory of components
+       BB.word16LE pageSizeNumElems
+    <> BB.word16LE pageSizeNumBlobs
+    <> BB.word16LE offKeyOffsets
+    <> BB.word16LE 0
+    -- (2) an array of 1-bit blob reference indicators
+    <> dfoldMap BB.word64LE (bmbits pageBlobRefBitmap)
+    -- (3) an array of 2-bit operation types
+    <> dfoldMap BB.word64LE (cmbits pageOperations)
+    -- (4) a pair of arrays of blob references
+    <> dfoldMap BB.word64LE pageBlobRefOffsets
+    <> dfoldMap BB.word32LE pageBlobRefSizes
+    -- (5) an array of key offsets
+    <> dfoldMap BB.word16LE pageKeyOffsets
+    -- (6) an array of value offsets
+    <> case pageValueOffsets of
+          Left   offsets -> dfoldMap BB.word16LE offsets
+          Right (offset1, offset2) -> BB.word16LE offset1
+                                  <> BB.word32LE offset2
+    -- (7) the concatenation of all keys
+    <> dfoldMap rawKey (unStricterList pageKeys)
+    -- (8) the concatenation of all values
+    <> dfoldMap rawValue (unStricterList pageValues)
```

Comment:
Just a nitpick, but `unStricterList` should not be necessary here I think, since you have `Foldable StricterList`.